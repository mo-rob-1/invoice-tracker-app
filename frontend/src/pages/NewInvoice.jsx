import { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";
import { toast } from "react-toastify";
import { createInvoice, reset } from "../features/invoices/invoiceSlice";
import BackButton from "../components/BackButton";

function NewInvoice() {
  const { user } = useSelector((state) => state.auth);
  const { isLoading, isError, isSuccess, message } = useSelector((state) => state.invoices);

  const [name] = useState(user?.name);
  const [email] = useState(user?.email);
  const [itemName, setItemName] = useState("");
  const [description, setDescription] = useState("");
  const [invoiceDate, setInvoiceDate] = useState("");
  const [userAddress, setUserAddress] = useState("");
  const [userCity, setUserCity] = useState("");
  const [userPostCode, setUserPostCode] = useState("");
  const [userCountry, setUserCountry] = useState("");
  const [quantity, setQuantity] = useState(0);
  const [paymentTerms, setPaymentTerms] = useState([]);
  const [clientName, setClientName] = useState("");
  const [clientEmail, setClientEmail] = useState("");
  const [clientAddress, setClientAddress] = useState("");
  const [clientPostCode, setClientPostCode] = useState("");
  const [clientCountry, setClientCountry] = useState("");
  const [clientCity, setClientCity] = useState("");
  const [price, setPrice] = useState(0);

  const dispatch = useDispatch();
  const navigate = useNavigate();

  useEffect(() => {
    if (isError) {
      toast.error(message);
    }

    if (isSuccess) {
      dispatch(reset());
      navigate("/invoices");
    }

    dispatch(reset());
  }, [dispatch, isError, isSuccess, navigate, message]);

  const onSubmit = (e) => {
    e.preventDefault();
    console.log(itemName);
    dispatch(
      createInvoice({
        itemName,
        description,
        invoiceDate,
        clientCity,
        quantity,
        userAddress,
        userCity,
        userPostCode,
        userCountry,
        clientName,
        clientEmail,
        clientAddress,
        clientPostCode,
        clientCountry,
        price,
        paymentTerms,
      })
    );
  };

  if (isLoading) {
    return <h3>Loading...</h3>;
  }

  return (
    <>
      <BackButton url="/" />
      <section className="heading">
        <h1>Create New Invoice</h1>
        <p>Please fill out the form below</p>
      </section>

      <section className="form">
        <div className="form-group">
          <label htmlFor="name" data-testid="seller-name-label">
            Seller's Name
          </label>
          <input type="text" className="form-control" value={name} disabled data-testid="seller-name-input" />
        </div>

        <div className="form-group">
          <label htmlFor="email" data-testid="seller-email-label">
            Seller's Email
          </label>
          <input type="text" className="form-control" value={email} disabled data-testid="seller-email-input" />
        </div>

        <form onSubmit={onSubmit} data-testid="invoice-form">
          <h3>Bill From</h3>
          <div className="form-group">
            <label htmlFor="userAddress" data-testid="street-address-label">
              Street Address
            </label>
            <input
              name="userAddress"
              id="userAddress"
              data-testid="street-address-input"
              value={userAddress}
              onChange={(e) => setUserAddress(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label htmlFor="userCity" data-testid="city-label">
              City
            </label>
            <input
              name="userCity"
              id="userCity"
              data-testid="city-input"
              value={userCity}
              onChange={(e) => setUserCity(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label htmlFor="userPostCode" data-testid="post-code-label">
              Post Code
            </label>
            <input
              name="userPostCode"
              id="userPostCode"
              data-testid="post-code-input"
              value={userPostCode}
              onChange={(e) => setUserPostCode(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label htmlFor="userCountry" data-testid="country-label">
              Country
            </label>
            <input
              name="userCountry"
              id="userCountry"
              data-testid="country-input"
              value={userCountry}
              onChange={(e) => setUserCountry(e.target.value)}
            />
          </div>
          <h3>Bill To</h3>
          <div className="form-group">
            <label htmlFor="clientName" data-testid="client-name-label">
              Client's Name
            </label>
            <input
              name="clientName"
              id="clientName"
              data-testid="client-name-input"
              value={clientName}
              onChange={(e) => setClientName(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label htmlFor="clientEmail" data-testid="client-email-label">
              Client's Email
            </label>
            <input
              name="clientEmail"
              id="clientEmail"
              data-testid="client-email-input"
              value={clientEmail}
              onChange={(e) => setClientEmail(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label htmlFor="clientAddress" data-testid="client-address-label">
              Street Address
            </label>
            <input
              name="clientAddress"
              id="clientAddress"
              data-testid="client-address-input"
              value={clientAddress}
              onChange={(e) => setClientAddress(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label htmlFor="clientCity" data-testid="client-city-label">
              City
            </label>
            <input
              name="clientCity"
              id="clientCity"
              data-testid="client-city-input"
              value={clientCity}
              onChange={(e) => setClientCity(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label htmlFor="clientPostCode" data-testid="client-post-code-label">
              Post Code
            </label>
            <input
              name="clientPostCode"
              id="clientPostCode"
              data-testid="client-post-code-input"
              value={clientPostCode}
              onChange={(e) => setClientPostCode(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label htmlFor="clientCountry" data-testid="client-country-label">
              Country
            </label>
            <input
              name="clientCountry"
              id="clientCountry"
              data-testid="client-country-input"
              value={clientCountry}
              onChange={(e) => setClientCountry(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label htmlFor="invoiceDate" data-testid="invoice-date-label">
              Invoice Date
            </label>
            <input
              type={"date"}
              name="invoiceDate"
              id="invoiceDate"
              data-testid="invoice-date-input"
              value={invoiceDate}
              onChange={(e) => setInvoiceDate(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label htmlFor="paymentTerms" data-testid="payment-terms-label">
              Payment Terms
            </label>
            <select
              name="paymentTerms"
              id="paymentTerms"
              data-testid="payment-terms-select"
              value={paymentTerms}
              onChange={(e) => setPaymentTerms(e.target.value)}
            >
              <option value="Net 1 Day">Net 1 Day</option>
              <option value="Net 7 Days">Net 7 Days</option>
              <option value="Net 14 Days">Net 14 Days</option>
              <option value="Net 30 Days">Net 30 Days</option>
            </select>
          </div>
          <h3>Invoice Summary</h3>
          <div className="form-group">
            <label htmlFor="itemName" data-testid="item-name-label">
              Item Name
            </label>
            <input
              name="itemName"
              id="itemName"
              data-testid="item-name-input"
              value={itemName}
              onChange={(e) => setItemName(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label htmlFor="description" data-testid="description-label">
              Description
            </label>
            <textarea
              name="description"
              id="description"
              data-testid="description-textarea"
              className="form-control"
              value={description}
              onChange={(e) => setDescription(e.target.value)}
            ></textarea>
          </div>
          <div className="form-group">
            <label htmlFor="quantity" data-testid="quantity-label">
              Quantity
            </label>
            <input
              name="quantity"
              id="quantity"
              data-testid="quantity-input"
              value={quantity}
              onChange={(e) => setQuantity(e.target.value)}
              type="number"
              step="1"
            />
          </div>
          <div className="form-group">
            <label htmlFor="price" data-testid="price-label">
              Price
            </label>
            <input
              name="price"
              id="price"
              data-testid="price-input"
              value={price}
              onChange={(e) => setPrice(e.target.value)}
            />
          </div>
          <div className="form-group">
            <button className="btn btn-block">Submit</button>
          </div>
          <div className="height"></div>
        </form>
      </section>
    </>
  );
}

export default NewInvoice;
