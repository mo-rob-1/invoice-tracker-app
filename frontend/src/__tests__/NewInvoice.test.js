import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Provider } from "react-redux";
import { store } from "../app/store";
import { BrowserRouter as Router } from "react-router-dom";
import NewInvoice from "../pages/NewInvoice";

describe("NewInvoice", () => {
  it("renders the new invoice page", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Provider store={store} history={history}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the form", () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
      <Provider store={store} history={history}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );
    expect(getByTestId("invoice-form")).toBeInTheDocument();
  });

  it("renders the texts", () => {
    const history = createMemoryHistory();
    const { getByText } = render(
      <Provider store={store} history={history}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    expect(getByText("Create New Invoice")).toBeInTheDocument();
    expect(getByText("Bill From")).toBeInTheDocument();
    expect(getByText("Bill To")).toBeInTheDocument();
    expect(getByText("Invoice Summary")).toBeInTheDocument();
  });

  it("renders the seller's name input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("seller-name-label");
    expect(label).toBeTruthy();

    const input = getByTestId("seller-name-input");
    expect(input).toBeTruthy();
  });

  it("renders the seller's email address input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("seller-email-label");
    expect(label).toBeTruthy();

    const input = getByTestId("seller-email-input");
    expect(input).toBeTruthy();
  });

  it("renders the street address input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("street-address-label");
    expect(label).toBeTruthy();

    const input = getByTestId("street-address-input");
    expect(input).toBeTruthy();
  });

  it("renders the city input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("city-label");
    expect(label).toBeTruthy();

    const input = getByTestId("city-input");
    expect(input).toBeTruthy();
  });

  it("renders the post code input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("post-code-label");
    expect(label).toBeTruthy();

    const input = getByTestId("post-code-input");
    expect(input).toBeTruthy();
  });

  it("renders the user country input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("country-label");
    expect(label).toBeTruthy();

    const input = getByTestId("country-input");
    expect(input).toBeTruthy();
  });

  it("renders the client name input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("client-name-label");
    expect(label).toBeTruthy();

    const input = getByTestId("client-name-input");
    expect(input).toBeTruthy();
  });

  it("renders the client email address input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("client-email-label");
    expect(label).toBeTruthy();

    const input = getByTestId("client-email-input");
    expect(input).toBeTruthy();
  });

  it("renders the client street address input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("client-address-label");
    expect(label).toBeTruthy();

    const input = getByTestId("client-address-input");
    expect(input).toBeTruthy();
  });

  it("renders the client city input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("client-city-label");
    expect(label).toBeTruthy();

    const input = getByTestId("client-city-input");
    expect(input).toBeTruthy();
  });

  it("renders the client post code input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("client-post-code-label");
    expect(label).toBeTruthy();

    const input = getByTestId("client-post-code-input");
    expect(input).toBeTruthy();
  });

  it("renders the client country input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("client-country-label");
    expect(label).toBeTruthy();

    const input = getByTestId("client-country-input");
    expect(input).toBeTruthy();
  });

  it("renders the invoice date input and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("invoice-date-label");
    expect(label).toBeTruthy();

    const input = getByTestId("invoice-date-input");
    expect(input).toBeTruthy();
  });

  it("renders the payment terms select and label", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("payment-terms-label");
    expect(label).toBeTruthy();

    const select = getByTestId("payment-terms-select");
    expect(select).toBeTruthy();
  });

  it("renders the item name label and input", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("item-name-label");
    expect(label).toBeTruthy();

    const input = getByTestId("item-name-input");
    expect(input).toBeTruthy();
  });

  it("renders the item description label and textarea", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("description-label");
    expect(label).toBeTruthy();

    const textarea = getByTestId("description-textarea");
    expect(textarea).toBeTruthy();
  });

  it("renders the item quantity label and input", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("quantity-label");
    expect(label).toBeTruthy();

    const input = getByTestId("quantity-input");
    expect(input).toBeTruthy();
  });

  it("renders the item price label and input", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const label = getByTestId("price-label");
    expect(label).toBeTruthy();

    const input = getByTestId("price-input");
    expect(input).toBeTruthy();
  });

  it("renders the submit button", () => {
    const { getByRole } = render(
      <Provider store={store}>
        <Router>
          <NewInvoice />
        </Router>
      </Provider>
    );

    const { getComputedStyle } = window;
    window.getComputedStyle = (elt) => getComputedStyle(elt);
    getByRole("button", { name: /submit/i });
  });
});
