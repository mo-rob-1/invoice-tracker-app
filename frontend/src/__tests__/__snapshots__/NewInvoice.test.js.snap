// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewInvoice renders the new invoice page 1`] = `
<DocumentFragment>
  <a
    class="btn btn-reverse btn-back"
    href="/"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 504C119 504 8 393 8 256S119 8 256 8s248 111 248 248-111 248-248 248zm28.9-143.6L209.4 288H392c13.3 0 24-10.7 24-24v-16c0-13.3-10.7-24-24-24H209.4l75.5-72.4c9.7-9.3 9.9-24.8.4-34.3l-11-10.9c-9.4-9.4-24.6-9.4-33.9 0L107.7 239c-9.4 9.4-9.4 24.6 0 33.9l132.7 132.7c9.4 9.4 24.6 9.4 33.9 0l11-10.9c9.5-9.5 9.3-25-.4-34.3z"
      />
    </svg>
     Back
  </a>
  <section
    class="heading"
  >
    <h1>
      Create New Invoice
    </h1>
    <p>
      Please fill out the form below
    </p>
  </section>
  <section
    class="form"
  >
    <div
      class="form-group"
    >
      <label
        data-testid="seller-name-label"
        for="name"
      >
        Seller's Name
      </label>
      <input
        class="form-control"
        data-testid="seller-name-input"
        disabled=""
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        data-testid="seller-email-label"
        for="email"
      >
        Seller's Email
      </label>
      <input
        class="form-control"
        data-testid="seller-email-input"
        disabled=""
        type="text"
        value=""
      />
    </div>
    <form
      data-testid="invoice-form"
    >
      <h3>
        Bill From
      </h3>
      <div
        class="form-group"
      >
        <label
          data-testid="street-address-label"
          for="userAddress"
        >
          Street Address
        </label>
        <input
          data-testid="street-address-input"
          id="userAddress"
          name="userAddress"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="city-label"
          for="userCity"
        >
          City
        </label>
        <input
          data-testid="city-input"
          id="userCity"
          name="userCity"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="post-code-label"
          for="userPostCode"
        >
          Post Code
        </label>
        <input
          data-testid="post-code-input"
          id="userPostCode"
          name="userPostCode"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="country-label"
          for="userCountry"
        >
          Country
        </label>
        <input
          data-testid="country-input"
          id="userCountry"
          name="userCountry"
          value=""
        />
      </div>
      <h3>
        Bill To
      </h3>
      <div
        class="form-group"
      >
        <label
          data-testid="client-name-label"
          for="clientName"
        >
          Client's Name
        </label>
        <input
          data-testid="client-name-input"
          id="clientName"
          name="clientName"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="client-email-label"
          for="clientEmail"
        >
          Client's Email
        </label>
        <input
          data-testid="client-email-input"
          id="clientEmail"
          name="clientEmail"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="client-address-label"
          for="clientAddress"
        >
          Street Address
        </label>
        <input
          data-testid="client-address-input"
          id="clientAddress"
          name="clientAddress"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="client-city-label"
          for="clientCity"
        >
          City
        </label>
        <input
          data-testid="client-city-input"
          id="clientCity"
          name="clientCity"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="client-post-code-label"
          for="clientPostCode"
        >
          Post Code
        </label>
        <input
          data-testid="client-post-code-input"
          id="clientPostCode"
          name="clientPostCode"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="client-country-label"
          for="clientCountry"
        >
          Country
        </label>
        <input
          data-testid="client-country-input"
          id="clientCountry"
          name="clientCountry"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="invoice-date-label"
          for="invoiceDate"
        >
          Invoice Date
        </label>
        <input
          data-testid="invoice-date-input"
          id="invoiceDate"
          name="invoiceDate"
          type="date"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="payment-terms-label"
          for="paymentTerms"
        >
          Payment Terms
        </label>
        <select
          data-testid="payment-terms-select"
          id="paymentTerms"
          name="paymentTerms"
        >
          <option
            value="Net 1 Day"
          >
            Net 1 Day
          </option>
          <option
            value="Net 7 Days"
          >
            Net 7 Days
          </option>
          <option
            value="Net 14 Days"
          >
            Net 14 Days
          </option>
          <option
            value="Net 30 Days"
          >
            Net 30 Days
          </option>
        </select>
      </div>
      <h3>
        Invoice Summary
      </h3>
      <div
        class="form-group"
      >
        <label
          data-testid="item-name-label"
          for="itemName"
        >
          Item Name
        </label>
        <input
          data-testid="item-name-input"
          id="itemName"
          name="itemName"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="description-label"
          for="description"
        >
          Description
        </label>
        <textarea
          class="form-control"
          data-testid="description-textarea"
          id="description"
          name="description"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="quantity-label"
          for="quantity"
        >
          Quantity
        </label>
        <input
          data-testid="quantity-input"
          id="quantity"
          name="quantity"
          step="1"
          type="number"
          value="0"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          data-testid="price-label"
          for="price"
        >
          Price
        </label>
        <input
          data-testid="price-input"
          id="price"
          name="price"
          value="0"
        />
      </div>
      <div
        class="form-group"
      >
        <button
          class="btn btn-block"
        >
          Submit
        </button>
      </div>
      <div
        class="height"
      />
    </form>
  </section>
</DocumentFragment>
`;
